app/vmalert/unittest.go

```diff
@@ -0,0 +1,780 @@
+		for _, g := range groups {
+			resolveDuration := getResolveDuration(g.Interval, *resendDelay, *maxResolveDuration)
+			errs := e.execConcurrently(context.Background(), g.Rules, ts, g.Concurrency, resolveDuration, g.Limit)
+			for err := range errs {
+				if err != nil {
+					checkErrs = append(checkErrs, fmt.Errorf("failed to exec group: %q, time: %s, err: %w", g.Name,
+						ts, err))
+				}
+			}
+			// need to flush immediately cause evaluation may rely on other group's result.
+			// we don't rely on remoteWrite.flushInterval to do this, cause it can't be controlled.
+			rw.DebugFlush()
+			vmstorage.Storage.DebugFlush()
+		}
+
+		// check alert at every eval time
+		for evalIndex < len(alertEvalTimes) {
+			if ts.Sub(testStartTime) > alertEvalTimes[evalIndex] ||
+				alertEvalTimes[evalIndex] >= ts.Add(evalInterval).Sub(testStartTime) {
+				break
+			}
+			gotAlertsMap := map[string]labelsAndAnnotations{}
+			for _, g := range groups {
+				for _, rule := range g.Rules {
+					ar, isAlertRule := rule.(*AlertingRule)
+					if !isAlertRule {
+						continue
+					}
+					if _, ok := alertExpResultMap[time.Duration(ts.UnixNano())][ar.Name]; ok {
+						for _, got := range ar.alerts {
+							if got.State != notifier.StateFiring {
+								continue
+							}
+							gotAlertsMap[ar.Name] = append(gotAlertsMap[ar.Name], labelAndAnnotation{
+								Labels:      convertToLabels(got.Labels),
+								Annotations: convertToLabels(got.Annotations),
+							})
+						}
+					}
+
+				}
+			}
+
+			for alertName, res := range alertExpResultMap[alertEvalTimes[evalIndex]] {
+				var expAlerts labelsAndAnnotations
+				for _, expAlert := range res {
+					if expAlert.ExpLabels == nil {
+						expAlert.ExpLabels = make(map[string]string)
+					}
+					// alertNameLabel is added as additional labels in vmalert.
+					expAlert.ExpLabels[alertNameLabel] = alertName
+					expAlerts = append(expAlerts, labelAndAnnotation{
+						Labels:      convertToLabels(expAlert.ExpLabels),
+						Annotations: convertToLabels(expAlert.ExpAnnotations),
+					})
+				}
+				sort.Sort(expAlerts)
+
+				gotAlerts := gotAlertsMap[alertName]
+				sort.Sort(gotAlerts)
+				if !reflect.DeepEqual(expAlerts, gotAlerts) {
+					var testName string
+					if tg.TestGroupName != "" {
+						testName = fmt.Sprintf("groupname: %s,\n", tg.TestGroupName)
+					}
+					expString := indentLines(expAlerts.String(), "            ")
+					gotString := indentLines(gotAlerts.String(), "            ")
+					checkErrs = append(checkErrs, fmt.Errorf("\n%s    alertname: %s, time: %s, \n        exp:%v, \n        got:%v ",
+						testName, alertName, alertEvalTimes[evalIndex].String(), expString, gotString))
+				}
+			}
+			evalIndex++
+		}
+
+	}
+
+	// check metricsql expr test
+	queries := q.BuildWithParams(datasource.QuerierParams{QueryParams: url.Values{"nocache": {"1"}, "latency_offset": {"1ms"}, "search.latencyOffset": {"1ms"}}, DataSourceType: "prometheus", Debug: true})
```

Comment:
why latecny offset is mentioned two times?
why debug is set to true?

Reply:
All for debug, remove dup now)